// Annotation-framework/src/lib/processors/inject/ComponentProcessor.js
// src/lib/processors/resources/ComponentProcessor.js

/**
 * @file ComponentProcessor.js
 * @description Processor for the @Component annotation.
 *
 * Responsibilities:
 *  - Handles class registration and instantiation
 *  - Supports all lifecycle stages: Scan, Instantiate, DI, Init, Run
 *  - Integrates with DI container and LifecycleManager
 */

const DiContainer = require('../../../core/DiContainer');
const LifecycleManager = require('../../../core/LifecycleManager');

class ComponentProcessor {
  /**
   * Called during the Scan stage.
   * Can validate metadata or prepare context.
   * @param {object} meta - Annotation metadata
   * @param {object} ctx - Stage context
   */
  processScan(meta, ctx) {
    if (Array.isArray(ctx.log)) {

      console.log("ComponentProcessor scan", meta.className);

      ctx.log.push(`@Component scanned: ${meta.className}`);
    }
  }

  /**
   * Called during the Instantiate stage.
   * Creates a singleton instance via LifecycleManager and registers it in the container.
   * @param {object} meta - Annotation metadata
   * @param {object} ctx - Stage context
   */
  processInstantiate(meta, ctx) {
    // Dynamically require the class at instantiation time to avoid circular dependency issues
    let classRef = meta.classRef;
    if (!classRef && meta.filePath && meta.className) {
      try {
        const required = require(meta.filePath);
        if (required && typeof required === 'function' && required.name === meta.className) {
          classRef = required;
        } else if (required && typeof required === 'object' && required[meta.className]) {
          classRef = required[meta.className];
        } else if (required && typeof required === 'function') {
          classRef = required;
        } else if (required && typeof required === 'object' && required.default && typeof required.default === 'function') {
          classRef = required.default;
        }
        if (!classRef && typeof required === 'function') {
          classRef = required;
        }
      } catch (e) {
        if (Array.isArray(ctx.log)) {
          ctx.log.push(`[ERROR] ComponentProcessor: Could not require class '${meta.className}' from '${meta.filePath}': ${e.message}`);
        }
      }
    }
    if (!classRef) {
      throw new TypeError(`ComponentProcessor: Could not resolve classRef for '${meta.className}' from '${meta.filePath}'`);
    }
    const instance = LifecycleManager.create(classRef, meta.className);
    meta.instance = instance;

    // Register in DI container if available
    if (ctx.container?.register) {
      ctx.container.register(meta.className, instance);
    }

    // Register as qualifier for property-level injection
    if (ctx.container?.registerQualifier) {
      ctx.container.registerQualifier(meta.className, instance);
    }

    if (Array.isArray(ctx.log)) {
      ctx.log.push(`@Component instantiated: ${meta.className}`);
    }
  }

  /**
   * Called during the DI stage.
   * Injects dependencies into the instance via DI container.
   * @param {object} meta - Annotation metadata
   * @param {object} ctx - Stage context
   */
  processDI(meta, ctx) {
    // If this is a property-level annotation (no meta.instance, but meta.className and ctx.container)
    if (!meta.instance && meta.className && ctx.container && typeof ctx.container.resolve === 'function') {
      try {
        meta.instance = ctx.container.resolve(meta.className);
      } catch (e) {
        // ignore if not found
      }
    }

    const instance = meta.instance;
    if (!instance) return;

    for (const [key, value] of DiContainer.instances.entries()) {
      if (key.startsWith('__inject:')) {
        const propName = key.split(':')[1];
        if (propName in instance) {
          instance[propName] = value;
          if (Array.isArray(ctx.log)) {
            ctx.log.push(`@Component DI: ${meta.className}.${propName} ← ${key}`);
          }
        }
      }
    }
  }

  /**
   * Called during the Init stage.
   * Executes any lifecycle initialization logic (PostConstruct etc.)
   * @param {object} meta - Annotation metadata
   * @param {object} ctx - Stage context
   */
  processInit(meta, ctx) {
    const instance = meta.instance;
    if (instance && typeof instance.initialize === 'function') {
      instance.initialize();
      if (Array.isArray(ctx.log)) {
        ctx.log.push(`@Component initialized: ${meta.className}`);
      }
    }
  }

  /**
   * Called during the Run stage.
   * Executes business logic or start hooks.
   * @param {object} meta - Annotation metadata
   * @param {object} ctx - Stage context
   */
  processRun(meta, ctx) {
    const instance = meta.instance;
    if (instance && typeof instance.run === 'function') {
      // Recursively ensure DI for all properties that are components
      for (const key of Object.keys(instance)) {
        const prop = instance[key];
        // If the property is an object and has a run method, assume it's a component
        if (prop && typeof prop === 'object' && typeof prop.run === 'function') {
          // Find the meta for this property in ctx.annotations
          const childMeta = ctx.annotations?.find(m => m.meta?.instance === prop);
          if (childMeta && childMeta.processor && typeof childMeta.processor.processDI === 'function') {
            childMeta.processor.processDI(childMeta.meta, ctx);
          }
        }
      }
      instance.run();
      if (Array.isArray(ctx.log)) {
        ctx.log.push(`@Component run executed: ${meta.className}`);
      }
    }
  }
}

module.exports = new ComponentProcessor;
