// Annotation-framework/src/lib/processors/inject/QualifierProcessor.js
/**
 * @file QualifierProcessor.js
 * @description Processor for the @Qualifier annotation.
 *
 * Responsibilities:
 *  - On property-level: injects a dependency resolved by qualifier into the annotated property
 *  - Does NOT register new instances
 *  - Supports all lifecycle stages (Scan, Instantiate, DI, Init, Run)
 */

class QualifierProcessor {
  processScan(meta, ctx) {
    // optional scan logic
  }

  processInstantiate(meta, ctx) {
    // optional instantiate logic
  }

  processDI(meta, ctx) {
    const [qualifierName] = meta.args || [];

    if (!qualifierName) {
      ctx.log?.push?.(`@Qualifier requires a name argument. Skipped on ${meta.className}`);
      console.log(`[DEBUG QualifierProcessor] No qualifierName for ${meta.className}.${meta.propertyName}`);
      return;
    }

    // Property-level injection
    if (meta.propertyName) {
      // Resolve parent instance if not set
      if (!meta.instance && ctx.container && typeof ctx.container.resolve === 'function' && meta.className) {
        try {
          meta.instance = ctx.container.resolve(meta.className);
          console.log(`[DEBUG QualifierProcessor] Resolved parent instance for ${meta.className}:`, meta.instance);
        } catch (e) {
          console.log(`[DEBUG QualifierProcessor] Failed to resolve parent instance for ${meta.className}:`, e);
        }
      }
      if (meta.instance && typeof ctx.container.resolveByQualifier === "function") {
        const dep = ctx.container.resolveByQualifier(qualifierName);
        console.log(`[DEBUG QualifierProcessor] Attempting to inject qualifier '${qualifierName}' into ${meta.className}.${meta.propertyName}. Dep:`, dep);
        if (dep) {
          meta.instance[meta.propertyName] = dep;
          console.log(`[DEBUG QualifierProcessor] Injected qualifier '${qualifierName}' into ${meta.className}.${meta.propertyName}`);
          ctx.log?.push?.(`@Qualifier injected: ${meta.propertyName} ← ${qualifierName}`)
        } else {
          console.log(`[DEBUG QualifierProcessor] Could not resolve dependency for qualifier '${qualifierName}'`);
          ctx.log?.push?.(`@Qualifier could not resolve dependency for '${qualifierName}'`)
        }
      } else {
        console.log(`[DEBUG QualifierProcessor] meta.instance is undefined for property-level injection: ${meta.className}.${meta.propertyName}`);
      }
    }
  }

  processInit(meta, ctx) {
    // optional init logic
  }

  processRun(meta, ctx) {
    // optional run logic
  }
}

// ⚡ EXPORT AS SINGLETON
module.exports = new QualifierProcessor();
