// Annotation-framework/src/configurations/framework.config.js
const path = require('path');

/**
 * Framework configuration.
 * This file is loaded during bootstrap and controls scanning, plugins, logging, lifecycle behavior, and stage execution.
 */
module.exports = {
  /**
   * Base directory of the project.
   * Used to resolve default scan paths.
   */
  baseDir: __dirname,

  /**
   * Directories to scan recursively for annotations.
   * If empty, defaults to `${baseDir}/src`.
   */
  scanDirs: [
    path.join(__dirname, '../modules'),
    path.join(__dirname, '../services')
    // You can add more directories here
  ],

  /**
   * Plugin configuration.
   * Plugins are loaded from these directories and must follow the expected structure.
   */
  plugins: {
    enabled: true,

    /**
     * Plugin search paths.
     * The default includes 'NodeJSAnnotations' inside node_modules.
     * You can add more paths to support custom or external plugins.
     */
    paths: [
      // Default plugin directory
      path.join(__dirname, '../../node_modules/NodeJSAnnotations'),

      // 🔧 Example: Custom plugin directory inside your project
      path.join(__dirname, '../plugins'),

      // 🧩 Example: External vendor plugins
      path.join(__dirname, '../../vendor-plugins'),

      // 🛠️ Example: Shared workspace plugins
      path.join(__dirname, '../../../shared/plugins')
    ]
  },

  /**
   * Logging configuration.
   */
  logging: {
    level: 'info', // 'debug' | 'info' | 'warn' | 'error'
    enabled: true
  },

  /**
   * Lifecycle hook configuration.
   */
  lifecycle: {
    enablePostConstruct: true,
    enablePreDestroy: false
  },

  /**
   * Stage configuration.
   * Defines the ordered list of lifecycle stages executed during bootstrap.
   * Plugins can register processors for any of these stages.
   */
  stages: ['Scan', 'Instantiate', 'DI', 'VI', 'Init', 'Run']
};
