// Annotation-framework/src/core/AnnotationLoader.js
const fs = require('fs');
const path = require('path');
const AnnotationRegistry = require('./AnnotationRegistry');

/**
 * Lädt Annotationen und Prozessoren aus dem lib-Verzeichnis.
 * Erwartet Struktur:
 *  annotations/**
 *  processors/**
 * Unterstützt Erweiterungsprozessoren, die ausschließlich vom PluginLoader registriert werden.
 */
class AnnotationLoader {
  constructor() {
    /** @type {string} */
    this.annotationRoot = path.join(__dirname, '../lib/annotations');
    /** @type {string} */
    this.processorRoot = path.join(__dirname, '../lib/processors');

    /** @type {Map<string, Function[]>} */
    this.extensions = new Map(); // Erweiterungsprozessoren
  }

  /**
   * Startet den Ladevorgang für alle Annotationen und Prozessoren.
   * Führt alle 'onLoad' Erweiterungen nach dem Laden aus.
   */
  load() {
    const annotationFiles = this._collectFiles(this.annotationRoot);

    annotationFiles.forEach((filePath) => {
      const relativePath = path.relative(this.annotationRoot, filePath);
      const name = path.basename(filePath, '.js');
      const annotationClass = require(filePath);

      const processorPath = path.join(this.processorRoot, relativePath.replace('.js', 'Processor.js'));

      if (fs.existsSync(processorPath)) {
        const processorClass = require(processorPath);
        AnnotationRegistry.register(name, annotationClass, processorClass);

        const onLoadExtensions = this.extensions.get('onLoad') || [];
        onLoadExtensions.forEach((fn) => fn(name, annotationClass, processorClass));
      }
    });
  }

  /**
   * Rekursive Dateisammlung aus einem Verzeichnis.
   * @param {string} dir - Startverzeichnis
   * @returns {string[]} - Liste aller .js-Dateien
   */
  _collectFiles(dir) {
    let results = [];

    fs.readdirSync(dir).forEach((entry) => {
      const fullPath = path.join(dir, entry);
      const stat = fs.statSync(fullPath);

      if (stat.isDirectory()) {
        results = results.concat(this._collectFiles(fullPath));
      } else if (stat.isFile() && fullPath.endsWith('.js')) {
        results.push(fullPath);
      }
    });

    return results;
  }

  /**
   * Wird ausschließlich vom PluginLoader verwendet, um Erweiterungsprozessoren zu registrieren.
   * @param {string} type - Erweiterungstyp (z. B. 'onLoad')
   * @param {Function} extensionFn - Erweiterungsfunktion
   */
  registerExtension(type, extensionFn) {
    if (!this.extensions.has(type)) {
      this.extensions.set(type, []);
    }
    this.extensions.get(type).push(extensionFn);
  }

  /**
   * Gibt alle Erweiterungen eines bestimmten Typs zurück.
   * @param {string} type
   * @returns {Function[]}
   */
  getExtensions(type) {
    return this.extensions.get(type) || [];
  }
}

module.exports = new AnnotationLoader();
