// Annotation-framework/src/core/AnnotationRegistry.js
/**
 * Verwaltet die Registrierung von Annotationen und zugehörigen Prozessoren.
 * Erweiterungsprozessoren werden ausschließlich über den PluginLoader eingebracht.
 */
class AnnotationRegistry {
  constructor() {
    /**
     * Haupt-Registry für Annotationen und ihre Prozessoren.
     * @type {Map<string, {annotationClass: Function, processorClass: Function}>}
     */
    this.registry = new Map();

    /**
     * Erweiterungsprozessoren, gruppiert nach Typ.
     * Diese werden ausschließlich vom PluginLoader registriert.
     * @type {Map<string, Function[]>}
     */
    this.extensions = new Map();
  }

  /**
   * Registriert eine Annotation und ihren Prozessor.
   * Führt alle Erweiterungsprozessoren vom Typ 'onRegister' aus.
   * @param {string} name - Name der Annotation
   * @param {Function} annotationClass - Annotation-Klasse
   * @param {Function} processorClass - Prozessor-Klasse
   */
  register(name, annotationClass, processorClass) {
    const entry = { annotationClass, processorClass };
    this.registry.set(name, entry);

    const onRegisterExtensions = this.extensions.get('onRegister') || [];
    onRegisterExtensions.forEach((fn) => fn(name, entry));
  }

  /**
   * Wird vom PluginLoader aufgerufen, um Erweiterungsprozessoren zu registrieren.
   * @param {string} type - Erweiterungstyp (z. B. 'onRegister')
   * @param {Function} extensionFn - Erweiterungsfunktion
   */
  registerExtension(type, extensionFn) {
    if (!this.extensions.has(type)) {
      this.extensions.set(type, []);
    }
    this.extensions.get(type).push(extensionFn);
  }

  /**
   * Gibt alle Erweiterungen eines bestimmten Typs zurück.
   * @param {string} type
   * @returns {Function[]}
   */
  getExtensions(type) {
    return this.extensions.get(type) || [];
  }

  /**
   * Gibt die Prozessor-Klasse für eine Annotation zurück.
   * @param {string} name
   * @returns {Function | undefined}
   */
  getProcessor(name) {
    return this.registry.get(name)?.processorClass;
  }

  /**
   * Gibt die Annotation-Klasse zurück.
   * @param {string} name
   * @returns {Function | undefined}
   */
  getAnnotation(name) {
    return this.registry.get(name)?.annotationClass;
  }

  /**
   * Gibt alle registrierten Annotationen zurück.
   * @returns {Array<[string, {annotationClass: Function, processorClass: Function}]>}
   */
  getAll() {
    return Array.from(this.registry.entries());
  }
}

module.exports = new AnnotationRegistry();
