// Annotation-framework/src/core/AstAnnotationParser.js
const fs = require('fs');
const path = require('path');
const acorn = require('acorn');
const walk = require('acorn-walk');
const AnnotationRegistry = require('./AnnotationRegistry');
const winston = require('winston');

// Winston Logger konfigurieren
const logger = winston.createLogger({
  level: 'debug',
  format: winston.format.printf(({ level, message }) => {
    const now = new Date();
    const timestamp = now.toTimeString().split(' ')[0]; // HH:mm:ss
    return `${timestamp}-${level.toUpperCase()}-${message}`;
  }),
  transports: [new winston.transports.Console()]
});

class AstAnnotationParser {
  constructor() {
    this.extensions = new Map();
  }

  parseFile(filePath, context) {
    const self = this;
    const content = fs.readFileSync(filePath, 'utf-8');
    const comments = [];

    const ast = acorn.parse(content, {
      ecmaVersion: 'latest',
      sourceType: 'module',
      locations: true,
      onComment: comments
    });

    if (context && !Array.isArray(context.annotations)) {
      context.annotations = [];
    }

    const collectedAnnotations = [];

    // --- Class/function Annotations ---
    comments
      .filter(c => {
        // Only process comments that are not immediately before a property definition
        // (property-level annotation comments are handled separately)
        const nextLine = c.loc.end.line + 1;
        let isPropertyComment = false;
        walk.simple(ast, {
          ClassDeclaration(node) {
            if (node.body && Array.isArray(node.body.body)) {
              node.body.body.forEach(element => {
                if ((element.type === 'PropertyDefinition' || element.type === 'ClassProperty') && element.loc.start.line === nextLine) {
                  isPropertyComment = true;
                }
              });
            }
          }
        });
        return c.type === 'Block' && c.value.includes('@') && !isPropertyComment;
      })
      .forEach(comment => {
        const lines = comment.value
          .split('\n')
          .map(l => l.trim().replace(/^\*/, '').trim())
          .filter(Boolean);

        lines.forEach(line => {
          if (!line.startsWith('@')) return;

          const match = line.match(/^@(\w+)(?:\((.*)\))?/);
          if (!match) return;

          const name = match[1];
          const rawParams = match[2] || '';
          const params = self._parseFlexibleParams(rawParams);

          const AnnotationClass = AnnotationRegistry.getAnnotation(name);
          const processor = AnnotationRegistry.getProcessor(name);

          if (AnnotationClass && processor) {
            const annotationInstance = new AnnotationClass(params);
            const target = self._findTarget(ast, comment.loc.end.line);

            logger.debug(
              `Found @${name} on target='${target.name}' in file='${filePath}'`
            );

            // Scan-Stage direkt ausführen
            if (typeof processor.processScan === 'function') {
              processor.processScan(annotationInstance, target);
              logger.debug(
                `${processor.constructor.name}: processScan executed for @${name} on target='${target.name}'`
              );
            }

            collectedAnnotations.push({ processor, annotationInstance, target, name, filePath, propertyName: undefined });

            if (context && Array.isArray(context.annotations)) {
              const meta = {
                annotation: annotationInstance,
                name,
                className: target.name,
                classRef: undefined,
                propertyName: undefined,
                args: annotationInstance.args || [annotationInstance.alias, annotationInstance.module],
                target,
                filePath
              };
              context.annotations.push({ processor, meta });
            }
          }
        });
      });

    // --- Property-level Annotations ---
    const commentMap = new Map();
    for (const c of comments) {
      if (c.type === 'Block' && c.value.includes('@')) {
        commentMap.set(c.loc.end.line, c);
      }
    }

    walk.simple(ast, {
      ClassDeclaration(node) {
        const className = node.id?.name || undefined;
        let classRef = undefined;
        try {
          const required = require(filePath);
          if (required && typeof required === 'function' && required.name === className) {
            classRef = required;
          } else if (required && typeof required === 'object' && required[className]) {
            classRef = required[className];
          }
        } catch (e) {}

        if (node.body && Array.isArray(node.body.body)) {
          node.body.body.forEach(element => {
            // Support both PropertyDefinition (ES2022+) and ClassProperty (older ASTs)
            if (element.type === 'PropertyDefinition' || element.type === 'ClassProperty') {
              const keyName = element.key?.name || element.key || undefined;
              const propLine = element.loc.start.line;
              const comment = commentMap.get(propLine - 1);
              if (comment && keyName) {
                const lines = comment.value
                  .split('\n')
                  .map(l => l.trim().replace(/^\*/, '').trim())
                  .filter(Boolean);

                lines.forEach(line => {
                  if (!line.startsWith('@')) return;
                  const match = line.match(/^@(\w+)(?:\((.*)\))?/);
                  if (!match) return;
                  const name = match[1];
                  const rawParams = match[2] || '';
                  const params = self._parseFlexibleParams(rawParams);

                  const AnnotationClass = AnnotationRegistry.getAnnotation(name);
                  const processor = AnnotationRegistry.getProcessor(name);
                  if (AnnotationClass && processor) {
                    const annotationInstance = new AnnotationClass(params);

                    logger.debug(
                      `Found @${name} on property='${keyName}' of class='${className}' in file='${filePath}'`
                    );

                    // Target korrekt auf Property setzen
                    const targetForProperty = { name: keyName, type: 'property', className: className, propertyName: keyName };

                    // Scan-Stage direkt ausführen
                    if (typeof processor.processScan === 'function') {
                      processor.processScan(annotationInstance, targetForProperty, keyName);
                      logger.debug(
                        `${processor.constructor.name}: processScan executed for @${name} on property='${keyName}' of class='${className}'`
                      );
                    }

                    // Set meta.instance to parent class instance if available in DI container
                    let parentInstance = undefined;
                    try {
                      if (className && context && context.container && typeof context.container.resolve === 'function') {
                        parentInstance = context.container.resolve(className);
                      }
                    } catch (e) {}

                    const meta = {
                      annotation: annotationInstance,
                      name,
                      className,
                      classRef,
                      propertyName: keyName,
                      args: params && Object.keys(params).length > 0 ? Object.values(params) : annotationInstance.args || [annotationInstance.name],
                      filePath,
                      instance: parentInstance
                    };
                    if (context && Array.isArray(context.annotations)) {
                      context.annotations.push({ processor, meta });
                    }

                    collectedAnnotations.push({
                      processor,
                      annotationInstance,
                      target: targetForProperty,
                      name,
                      filePath,
                      propertyName: keyName
                    });
                  }
                });
              }
            }
          });
        }
      }
    });

    // Für andere Stages werden die Annotationen gesammelt, aber nicht direkt ausgeführt
    return collectedAnnotations;
  }

  _parseFlexibleParams(raw) {
    const params = {};
    if (!raw) return params;
    const parts = raw.match(/(?:[^\s,"]+|"[^"]*"|'[^']*')+/g) || [];
    parts.forEach((part, index) => {
      part = part.trim();
      const eqIndex = part.indexOf('=');
      if (eqIndex > -1) {
        const key = part.slice(0, eqIndex).trim();
        let value = part.slice(eqIndex + 1).trim();
        if (
          (value.startsWith('"') && value.endsWith('"')) ||
          (value.startsWith("'") && value.endsWith("'"))
        ) {
          value = value.slice(1, -1);
        }
        params[key] = value || undefined;
      } else {
        params[`param${index + 1}`] = part.replace(/^['"]|['"]$/g, '');
      }
    });
    return params;
  }

  _findTarget(ast, lineAfterComment) {
    let target = { name: 'UnknownTarget', type: 'unknown' };
    walk.simple(ast, {
      ClassDeclaration(node) {
        if (node.loc.start.line === lineAfterComment + 1) {
          target = { name: node.id.name, type: 'class' };
        }
      },
      FunctionDeclaration(node) {
        if (node.loc.start.line === lineAfterComment + 1) {
          target = { name: node.id.name, type: 'function' };
        }
      }
    });
    return target;
  }

  registerExtension(type, extensionFn) {
    if (!this.extensions.has(type)) {
      this.extensions.set(type, []);
    }
    this.extensions.get(type).push(extensionFn);
  }

  getExtensions(type) {
    return this.extensions.get(type) || [];
  }
}

module.exports = new AstAnnotationParser();
