// Annotation-framework/src/core/Bootstrap.js
const path = require('path');
const fs = require('fs');
const winston = require('winston');
const PluginLoader = require('./PluginLoader');
const AnnotationLoader = require('./AnnotationLoader');
const AnnotationRegistry = require('./AnnotationRegistry');
const AnnotationParser = require('./AstAnnotationParser');
const StageExecutor = require('./StageExecutor');
const { STAGES } = require('./StageRegistry');

// Default Winston Logger
const defaultLogger = winston.createLogger({
  level: 'info',
  format: winston.format.combine(
    winston.format.colorize(),
    winston.format.timestamp({ format: 'HH:mm:ss' }),
    winston.format.printf(
      ({ timestamp, level, message }) => `[${timestamp}] ${level}: ${message}`
    )
  ),
  transports: [new winston.transports.Console()]
});

class Bootstrap {
  static async run(options = {}) {
    const logger = options.logger || defaultLogger;

    logger.info('🚀 Starting framework...');

    logger.info('📦 Loading local annotations...');
    AnnotationLoader.load();

    logger.info('🔌 Loading plugins...');
    PluginLoader.load();

    logger.info('🧠 Registered annotations:');
    AnnotationRegistry.getAll().forEach(([name, { annotationClass, processorClass }]) => {
      logger.info(`- ${name}: ${annotationClass.name}, Processor: ${processorClass.constructor.name}`);
    });

    // Bestimme Scan-Verzeichnisse
    const baseDir = options.baseDir || path.resolve(__dirname, '..');
    const defaultScanDir = path.join(baseDir, 'src');
    const scanDirs = options.scanDirs?.length ? options.scanDirs : [defaultScanDir];

    // --- Scan-Phase vor allen Prozessoren ---
    logger.info('🔍 Starting annotation scan (Scan phase)...');
    const scanContext = {
      config: options,
      container: options.container || {},
      metadata: options.metadata || {},
      log: options.log || []
    };

    // Collect annotation meta for all files
    scanContext.annotations = [];

    scanDirs.forEach((dir) => {
      const resolvedDir = path.resolve(dir);
      logger.info(`📁 Scanning directory: ${resolvedDir}`);
      Bootstrap._scanDirectoryRecursive(resolvedDir, logger, scanContext);
    });

    // Nach Scan-Phase: Stage-Prozessoren ausführen
    logger.info('⚙️ Executing lifecycle stages...');
    for (const stage of STAGES) {
      logger.info(`➡️ Stage: ${stage}`);
      await StageExecutor.executeStage(stage, scanContext);
    }

    logger.info('✅ Framework fully initialized.');
  }

  static _scanDirectoryRecursive(dir, logger = defaultLogger, context) {
    if (!fs.existsSync(dir)) return;

    const entries = fs.readdirSync(dir, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);

      // System-Ordner ignorieren
      if (
        entry.name === 'node_modules' ||
        entry.name === '.git' ||
        entry.name === '.vscode' ||
        entry.name.startsWith('.')
      ) continue;

      if (entry.isDirectory()) {
        Bootstrap._scanDirectoryRecursive(fullPath, logger, context);
      } else if (entry.isFile() && fullPath.endsWith('.js')) {
        logger.debug(`🔎 Parsing file: ${fullPath}`);
        AnnotationParser.parseFile(fullPath, context);
      }
    }
  }
}

module.exports = Bootstrap;
