// Annotation-framework/src/core/LifecycleManager.js
const DiContainer = require('./DiContainer');

/**
 * The LifecycleManager orchestrates the instantiation and dependency injection of components.
 * All created instances are treated as singletons and stored in the DI container.
 * Supports plugin-based lifecycle extensions (registered via PluginLoader).
 */
class LifecycleManager {
  constructor() {
    /**
     * Extension processors grouped by type.
     * These are registered exclusively via PluginLoader.
     * @type {Map<string, Function[]>}
     */
    this.extensions = new Map();
  }

  /**
   * Creates or retrieves a singleton instance of the given class.
   *
   * This method performs:
   * - Checks if an instance with the given name already exists in the DI container
   * - If not, creates a new instance and injects dependencies
   * - Supports property and setter injection based on DI container keys
   * - Executes 'onCreate' lifecycle extensions
   * - Registers the instance in the DI container
   *
   * @param {Function} ClassRef - The class constructor to instantiate
   * @param {string} name - The unique name under which the instance is stored
   * @returns {Object} - The singleton instance
   */
  create(ClassRef, name) {
    if (DiContainer.has(name)) {
      return DiContainer.resolve(name);
    }

    const instance = new ClassRef();

    // Property injection
    for (const [key, value] of DiContainer.instances.entries()) {
      if (key.startsWith('__inject:')) {
        const propName = key.split(':')[1];
        if (propName in instance) {
          instance[propName] = value;
        }
      }
    }

    // Setter injection
    for (const [key, value] of DiContainer.instances.entries()) {
      if (key.startsWith('__injectMethod:')) {
        const methodName = key.split(':')[1];
        if (typeof instance[methodName] === 'function') {
          instance[methodName](value);
        }
      }
    }

    // Execute lifecycle extensions
    const onCreateHooks = this.extensions.get('onCreate') || [];
    onCreateHooks.forEach((fn) => fn(name, instance, ClassRef));

    DiContainer.set(name, instance);
    return instance;
  }

  /**
   * Initializes all services in the registry as singletons.
   * Executes 'onInitialize' lifecycle extensions.
   *
   * @param {Map<string, Function>} registry - A map of service names to class constructors
   */
  initializeAll(registry) {
    const onInitializeHooks = this.extensions.get('onInitialize') || [];
    onInitializeHooks.forEach((fn) => fn(registry));

    for (const [name, ClassRef] of registry.entries()) {
      this.create(ClassRef, name);
    }
  }

  /**
   * Registers an extension processor for a given lifecycle phase.
   * Only the PluginLoader should call this method.
   *
   * @param {string} type - Extension type (e.g. 'onCreate', 'onInitialize')
   * @param {Function} extensionFn - Extension function
   */
  registerExtension(type, extensionFn) {
    if (!this.extensions.has(type)) {
      this.extensions.set(type, []);
    }
    this.extensions.get(type).push(extensionFn);
  }

  /**
   * Returns all registered extensions of a given type.
   * @param {string} type
   * @returns {Function[]}
   */
  getExtensions(type) {
    return this.extensions.get(type) || [];
  }
}

module.exports = new LifecycleManager();
