// Annotation-framework/src/core/PluginLoader.js
const fs = require('fs');
const path = require('path');
const AnnotationRegistry = require('./AnnotationRegistry');
const DIContainer = require('./DiContainer');
const LifecycleManager = require('./LifecycleManager');
const AnnotationLoader = require('./AnnotationLoader');
const AstAnnotationParser = require('./AstAnnotationParser');
const config = require('../configurations/framework.config');
const StageExecutor = require('./StageExecutor'); // ✅ NEU: Stage-Verwaltung
const { STAGES } = require('./StageRegistry');    // ✅ NEU: zentrale Stage-Definition

/**
 * Loads plugins from configured directories and registers their annotations,
 * extension processors, and stage processors.
 * Supports plugin-based extensions for all core components, including the PluginLoader itself.
 */
class PluginLoader {
  constructor() {
    /** @type {string[]} */
    this.pluginDirs = config.plugins?.paths || [];

    /** @type {Map<string, Function[]>} */
    this.extensions = new Map(); // Erweiterungen für PluginLoader selbst

    /** @type {Record<string, any>} */
    this.extensionTargets = {
      AnnotationRegistry,
      DIContainer,
      LifecycleManager,
      AnnotationLoader,
      AstAnnotationParser,
      PluginLoader: this // ✅ NEU: PluginLoader als Erweiterungsziel
    };
  }

  /**
   * Loads all plugins from configured directories.
   *
   * For each plugin:
   * - Reads its package.json to determine load order
   * - Loads annotation classes from 'annotations' folder
   * - Loads corresponding processor classes from 'processors' folder
   * - Registers each annotation-processor pair in the AnnotationRegistry
   * - Loads and applies extension processors defined in package.json
   * - Registers stage processors if defined
   * - Executes 'onPluginLoaded' hooks
   */
  load() {
    const allPlugins = [];

    this.pluginDirs.forEach((pluginRoot) => {
      if (!fs.existsSync(pluginRoot)) return;

      const pluginFolders = fs.readdirSync(pluginRoot)
        .map((pluginName) => {
          const pluginPath = path.join(pluginRoot, pluginName);
          const pkgPath = path.join(pluginPath, 'package.json');

          if (!fs.existsSync(pkgPath)) return null;

          const pkg = require(pkgPath);
          const order = pkg.annotationLoadOrder || 1000;

          return { pluginPath, pkg, order };
        })
        .filter(Boolean);

      allPlugins.push(...pluginFolders);
    });

    // Sort all plugins by load order
    allPlugins.sort((a, b) => a.order - b.order);

    // Load annotations, processors, extensions, and stage processors
    allPlugins.forEach(({ pluginPath, pkg }) => {
      this.loadAnnotations(pluginPath);
      this.loadExtensions(pluginPath, pkg);
      this.loadStageProcessors(pluginPath); // ✅ NEU: Stage-Prozessoren laden

      // Execute PluginLoader lifecycle hook
      const onPluginLoaded = this.extensions.get('onPluginLoaded') || [];
      onPluginLoaded.forEach(fn => fn(pluginPath, pkg));
    });
  }

  /**
   * Loads and registers annotation classes and their processors.
   * @param {string} pluginPath
   */
  loadAnnotations(pluginPath) {
    const annotationDir = path.join(pluginPath, 'annotations');
    const processorDir = path.join(pluginPath, 'processors');

    if (!fs.existsSync(annotationDir) || !fs.existsSync(processorDir)) return;

    const annotationFiles = fs.readdirSync(annotationDir);

    annotationFiles.forEach((file) => {
      const name = path.basename(file, '.js');
      const annotationPath = path.join(annotationDir, file);
      const processorPath = path.join(processorDir, `${name}Processor.js`);

      if (fs.existsSync(processorPath)) {
        const AnnotationClass = require(annotationPath);
        const ProcessorClass = require(processorPath);
        AnnotationRegistry.register(name, AnnotationClass, ProcessorClass);
      }
    });
  }

  /**
   * Loads and applies extension processors defined in plugin's package.json.
   * @param {string} pluginPath
   * @param {Record<string, any>} pkg
   */
  loadExtensions(pluginPath, pkg) {
    const extensions = pkg.extensionProcessors || {};

    Object.entries(extensions).forEach(([targetName, processorPaths]) => {
      const target = this.extensionTargets[targetName];
      if (!target) return;

      processorPaths.forEach((relativePath) => {
        const fullPath = path.resolve(pluginPath, relativePath);
        if (!fs.existsSync(fullPath)) return;

        const Extension = require(fullPath);
        const instance = typeof Extension === 'function' ? new Extension() : Extension;

        if (typeof instance.apply === 'function') {
          instance.apply(target);
        } else if (typeof instance === 'function' && typeof target.registerExtension === 'function') {
          target.registerExtension('plugin', instance);
        }
      });
    });
  }

  /**
   * Loads and registers stage processors defined in plugin's index.js.
   * @param {string} pluginPath
   */
  loadStageProcessors(pluginPath) {
    const pluginEntry = path.join(pluginPath, 'index.js');
    if (!fs.existsSync(pluginEntry)) return;

    const plugin = require(pluginEntry);
    if (!plugin.processors) return;

    STAGES.forEach(stage => {
      const stageFns = plugin.processors[stage];
      if (Array.isArray(stageFns)) {
        stageFns.forEach(fn => {
          StageExecutor.registerStageProcessor(stage, fn);
        });
      }
    });
  }

  /**
   * Registers an extension processor for the PluginLoader itself.
   * Only the PluginLoader should call this method.
   * @param {string} type - Extension type (e.g. 'onPluginLoaded')
   * @param {Function} extensionFn
   */
  registerExtension(type, extensionFn) {
    if (!this.extensions.has(type)) {
      this.extensions.set(type, []);
    }
    this.extensions.get(type).push(extensionFn);
  }

  /**
   * Returns all registered extensions of a given type.
   * @param {string} type
   * @returns {Function[]}
   */
  getExtensions(type) {
    return this.extensions.get(type) || [];
  }
}

module.exports = new PluginLoader();
